control_panel_api/permissions.py

```diff
@@ -93,24 +95,35 @@ def has_object_permission(self, request, view, obj):
 
 
 class UserS3BucketPermissions(BasePermission):
-    """
-    - Superusers can do anything
-    - Normal users can list/retrieve UserS3Buckets
-
-    NOTE: Filters are applied before permissions
-    """
-
     def has_permission(self, request, view):
         if is_superuser(request.user):
             return True
 
         if request.user.is_anonymous():
             return False
 
-        return view.action in ('list', 'retrieve')
+        if view.action != 'create':
+            return True
 
+        serializer = UserS3BucketSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
 
-class ToolDeploymentPermissions(BasePermission):
+        accessible_buckets = S3Bucket.objects.administered_by(request.user)
+
+        if serializer.validated_data['s3bucket'] in accessible_buckets:
```

Comment:
We could avoid importing/using the serialiser at all by changing the `if` to query for existence, e.g.:

```python
# "ugly" proof of concept
if accessible_buckets.filter(pk=request.data['s3bucket']).exists():
    return True
```

Making a query has also the benefit of being more efficient for users with access to a large number of S3 buckets (maybe add something like that to to the `S3BucketQuerySet`? I don't know)